sdks/java/core/src/main/java/org/apache/beam/sdk/util/IOChannelUtils.java

```diff
@@ -51,18 +65,121 @@
   }
 
   /**
-   * Registers standard factories globally. This requires {@link PipelineOptions}
-   * to provide, e.g., credentials for GCS.
+   * Deregisters the scheme and the associated {@link IOChannelFactory}.
+   */
+  @VisibleForTesting
+  static void deregisterScheme(String scheme) {
+    FACTORY_MAP.remove(scheme);
+  }
+
+  /**
+   * Registers standard factories globally.
+   *
+   * <p>{@link PipelineOptions} are required to provide dependencies and
+   * pipeline level configuration to the individual {@link IOChannelFactory IOChannelFactories}.
+   *
+   * @deprecated use {@link #registerIOFactories}.
    */
+  @Deprecated
   public static void registerStandardIOFactories(PipelineOptions options) {
-    setIOFactory("gs", new GcsIOChannelFactory(options.as(GcsOptions.class)));
-    setIOFactory("file", new FileIOChannelFactory());
+    registerIOFactoriesAllowOverride(options);
+  }
+
+  /**
+   * Registers all {@link IOChannelFactory IOChannelFactories} from {@link ServiceLoader}.
+   *
+   * <p>{@link PipelineOptions} are required to provide dependencies and
+   * pipeline level configuration to the individual {@link IOChannelFactory IOChannelFactories}.
+   *
+   * <p>Multiple {@link IOChannelFactory IOChannelFactories} for the same scheme are not allowed.
+   *
+   * @throws IllegalStateException if multiple {@link IOChannelFactory IOChannelFactories}
+   * for the same scheme are detected.
+   */
+  public static void registerIOFactories(PipelineOptions options) {
+    registerIOFactoriesInternal(options, false /* override */);
+  }
+
+  /**
+   * Registers all {@link IOChannelFactory IOChannelFactories} from {@link ServiceLoader}.
+   *
+   * <p>This requires {@link PipelineOptions} to provide, e.g., credentials for GCS.
+   *
+   * <p>Override existing schemes is allowed.
+   *
+   * @deprecated This is currently to provide different configurations for tests and
+   * is still public for IOChannelFactory redesign purposes.
+   */
+  @Deprecated
+  @VisibleForTesting
+  public static void registerIOFactoriesAllowOverride(PipelineOptions options) {
+    registerIOFactoriesInternal(options, true /* override */);
+  }
+
+  private static void registerIOFactoriesInternal(
+      PipelineOptions options, boolean override) {
+    Set<IOChannelFactoryRegistrar> registrars =
+        Sets.newTreeSet(ReflectHelpers.ObjectsClassComparator.INSTANCE);
+    registrars.addAll(Lists.newArrayList(
+        ServiceLoader.load(IOChannelFactoryRegistrar.class, CLASS_LOADER)));
+
+    checkDuplicateScheme(registrars);
+
+    for (IOChannelFactoryRegistrar registrar : registrars) {
+      setIOFactoryInternal(
+          registrar.getScheme(),
+          registrar.fromOptions(options),
+          override);
+    }
+  }
+
+  @VisibleForTesting
+  static void checkDuplicateScheme(Set<IOChannelFactoryRegistrar> registrars) {
+    Multimap<String, IOChannelFactoryRegistrar> registrarsBySchemes = HashMultimap.create();
```

Comment:
I said to sort on the scheme which is a string, not on the IOChannelFactoryRegistrar.
Also, now that you mention it, sorting the values for the string based upon class name would give a stable sort.